javalanglib/src/main/scala/java/lang/Double.scala

```diff
@@ -63,22 +63,147 @@ object Double {
 
   @inline def valueOf(s: String): Double = valueOf(parseDouble(s))
 
-  private[this] lazy val doubleStrPat = new js.RegExp("^" +
-      "[\\x00-\\x20]*"   + // optional whitespace
-      "[+-]?"            + // optional sign
-      "(NaN|Infinity|"   + // special cases
-       "(\\d+\\.?\\d*|"  + // literal w/  leading digit
-        "\\.\\d+)"       + // literal w/o leading digit
-       "([eE][+-]?\\d+)?"+ // optional exponent
-      ")[fFdD]?"         + // optional float / double specifier (ignored)
-      "[\\x00-\\x20]*"   + // optional whitespace
+  private[this] lazy val doubleStrPat = new js.RegExp(
+      "^"                   +
+      "[\\x00-\\x20]*("     + // optional whitespace
+      "[+-]?"               + // optional sign
+      "(?:NaN|Infinity|"    + // special cases
+       "(?:\\d+\\.?\\d*|"   + // literal w/  leading digit
+        "\\.\\d+)"          + // literal w/o leading digit
+       "(?:[eE][+-]?\\d+)?" + // optional exponent
+      ")[fFdD]?"            + // optional float / double specifier (ignored)
+      ")[\\x00-\\x20]*"     + // optional whitespace
+      "$")
+
+  private[this] lazy val doubleStrHexPat = new js.RegExp(
+      "^"                   +
+      "[\\x00-\\x20]*"      + // optional whitespace
+      "([+-]?)"             + // optional sign
+      "0[xX]"               + // hex marker
+      "([0-9A-Fa-f]*)"      + // integral part
+      "\\.?([0-9A-Fa-f]*)"  + // fractional part
+      "[pP]([+-]?\\d+)"     + // binary exponent
+      "[fFdD]?"             + // optional float / double specifier (ignored)
+      "[\\x00-\\x20]*"      + // optional whitespace
       "$")
 
   def parseDouble(s: String): scala.Double = {
-    if (doubleStrPat.test(s))
-      js.Dynamic.global.parseFloat(s).asInstanceOf[scala.Double]
-    else
+    def fail(): Nothing =
       throw new NumberFormatException(s"""For input string: "$s"""")
+
+    // (Very) slow path for hexadecimal notation
+    def parseHexDoubleImpl(match2: js.RegExp.ExecResult): scala.Double = {
+      // scalastyle:off return
+
+      val signStr = match2(1).asInstanceOf[String]
+      val integralPartStr = match2(2).asInstanceOf[String]
+      val fractionalPartStr = match2(3).asInstanceOf[String]
+      val binaryExpStr = match2(4).asInstanceOf[String]
+
+      if (integralPartStr == "" && fractionalPartStr == "")
+        fail()
+
+      def overflow(): scala.Double =
+        if (signStr == "-") scala.Double.NegativeInfinity
+        else scala.Double.PositiveInfinity
+
+      def underflow(): scala.Double =
+        if (signStr == "-") -0.0
+        else 0.0
```

Comment:
Consider inlining this at its sole usage site. `underflow` is arguably a strange name for that usage, since the original string was basically a form of "0" (and *not* something with a too small fractional part).